### PR TITLE
Implementation of the shadow Thing feature on the device

### PR DESCRIPTION
**Pull request by ilcato**
Device implementation of shadow Thing feature.

The ArduinoIoTCloud update accept two new optional parameters:

void update(int mode = PROPERTIES_SYNC_FORCE_DEVICE, void (*callback)(void) = NULL);

the mode maybe:
  PROPERTIES_SYNC_FORCE_DEVICE, // default mode, after a connection/reconnection the device value of a property is forced on the cloud
  PROPERTIES_SYNC_FORCE_CLOUD, // after a connection/reconnection the cloud value of a property is forced on the device
  PROPERTIES_SYNC_AUTO, // after a connection/reconnection the timestamps of the changes on device and on cloud are used to choose which value the property will assume
  PROPERTIES_SYNC_CUSTOM // not yet implemented 

The callback is called after the processing.

### CODE REVIEW

**Comment by lxrobotics** [comment]
Could you please provide some unit tests for your changes?

File: src/ArduinoIoTCloud.cpp [hunk 1]
```diff
@@ -218,11 +251,47 @@ void ArduinoIoTCloudClass::update(int const reconnectionMaxRetries, int const re
...
-  if (length > 0) {
-    writeProperties(data, length);
+  if(!_propertiesSynchronized) {
+    if (millis() - _lastSyncRequestTickTime > TIMEOUT_FOR_LASTVALUES_SYNC) {
+      _firstReconnectionUpdate = true;
+      _propertiesSynchronized = true;
+    }
+  } 
+
+  if(((_mode == PROPERTIES_SYNC_FORCE_CLOUD) || (_mode == PROPERTIES_SYNC_AUTO)) && _firstReconnectionUpdate){
```
Comment (per1234) [inline src/ArduinoIoTCloud.cpp]:
```suggestion
  if(((_mode == PROPERTIES_SYNC_FORCE_CLOUD) || (_mode == PROPERTIES_SYNC_AUTO)) && _firstReconnectionUpdate) {
```

File: src/ArduinoIoTCloud.cpp [hunk 2]
```diff
@@ -218,11 +251,47 @@ void ArduinoIoTCloudClass::update(int const reconnectionMaxRetries, int const re
...
+    // get the last values of the shadow_properties from MQTT broker
+    getLastValues();
+    _lastSyncRequestTickTime = millis();
+    // properties will be not synchronized until the OnMessage callback will be activated on _shadowTopicIn
+    _firstReconnectionUpdate = false;
+    _propertiesSynchronized = false;
+    // store callback pointer to the function to be called after the sync process end
+    _syncCallback = callback;
+  }
```
Comment (per1234) [inline src/ArduinoIoTCloud.cpp]:
```suggestion
  }
```

File: src/ArduinoIoTCloud.cpp [hunk 3]
```diff
@@ -218,11 +251,47 @@ void ArduinoIoTCloudClass::update(int const reconnectionMaxRetries, int const re
...
+    getLastValues();
+    _lastSyncRequestTickTime = millis();
+    // properties will be not synchronized until the OnMessage callback will be activated on _shadowTopicIn
+    _firstReconnectionUpdate = false;
+    _propertiesSynchronized = false;
+    // store callback pointer to the function to be called after the sync process end
+    _syncCallback = callback;
+  } 
+
+  if(_mode == PROPERTIES_SYNC_FORCE_DEVICE){
```
Comment (per1234) [inline src/ArduinoIoTCloud.cpp]:
```suggestion
  if(_mode == PROPERTIES_SYNC_FORCE_DEVICE) {
```

File: src/ArduinoIoTCloud.cpp [hunk 4]
```diff
@@ -218,11 +251,47 @@ void ArduinoIoTCloudClass::update(int const reconnectionMaxRetries, int const re
...
+    _firstReconnectionUpdate = false;
+  }
```
Comment (per1234) [inline src/ArduinoIoTCloud.cpp]:
```suggestion
  }
```

File: src/ArduinoIoTCloud.cpp [hunk 5]
```diff
@@ -218,11 +251,47 @@ void ArduinoIoTCloudClass::update(int const reconnectionMaxRetries, int const re
...
+    // store callback pointer to the function to be called after the sync process end
+    _syncCallback = callback;
+  } 
+
+  if(_mode == PROPERTIES_SYNC_FORCE_DEVICE){
+    _firstReconnectionUpdate = false;
+  } 
+
+  // If in the properties are synchronized or the synch mode is PROPERTIES_SYNC_FORCE_DEVICE, send values to MQTT broker
+  if((_propertiesSynchronized) && !(_mode == CLOUDSERIAL_SYNC)){
```
Comment (per1234) [inline src/ArduinoIoTCloud.cpp]:
```suggestion
  if((_propertiesSynchronized) && !(_mode == CLOUDSERIAL_SYNC)) {
```

File: src/ArduinoIoTCloud.cpp [hunk 6]
```diff
@@ -296,6 +382,23 @@ void ArduinoIoTCloudClass::handleMessage(int length)
...
+    // set the flag to let know the update cycle that the properties are synchronized and the sync have been made followinf the _mode rules
+    _propertiesSynchronized = true;
+    // delay execution of LastValueCallback at the end of the update() cycle
+    _callGetLastValueCallback = true;
+  }
+}
+
+void ArduinoIoTCloudClass::getLastValues()
+{
+    uint8_t data[MQTT_TRANSMIT_BUFFER_SIZE];
```
Comment (per1234) [inline src/ArduinoIoTCloud.cpp]:
```suggestion
  uint8_t data[MQTT_TRANSMIT_BUFFER_SIZE];
```

File: src/ArduinoIoTCloud.cpp [hunk 7]
```diff
@@ -296,6 +382,23 @@ void ArduinoIoTCloudClass::handleMessage(int length)
...
+    int const length = Thing.getLastValues(data, sizeof(data));
```
Comment (per1234) [inline src/ArduinoIoTCloud.cpp]:
```suggestion
  int const length = Thing.getLastValues(data, sizeof(data));
```

File: src/ArduinoIoTCloud.cpp [hunk 8]
```diff
@@ -296,6 +382,23 @@ void ArduinoIoTCloudClass::handleMessage(int length)
...
+    if (length > 0) {
```
Comment (per1234) [inline src/ArduinoIoTCloud.cpp]:
```suggestion
  if (length > 0) {
```

File: src/ArduinoIoTCloud.cpp [hunk 9]
```diff
@@ -296,6 +382,23 @@ void ArduinoIoTCloudClass::handleMessage(int length)
...
+      writeShadowOut(data, length);
```
Comment (per1234) [inline src/ArduinoIoTCloud.cpp]:
```suggestion
    writeShadowOut(data, length);
```

File: src/ArduinoIoTCloud.cpp [hunk 10]
```diff
@@ -296,6 +382,23 @@ void ArduinoIoTCloudClass::handleMessage(int length)
...
+    }
```
Comment (per1234) [inline src/ArduinoIoTCloud.cpp]:
```suggestion
  }
```

**Pull request closed without merging**